// step-common/check.js
import { createOctokit } from "./github.js";

const MAX_ANNOTATIONS = 50;

export function splitAnnotations(annotations = []) {
  if (annotations.length <= MAX_ANNOTATIONS) {
    return {
      annotations,
      truncated: false,
      restCount: 0,
    };
  }

  return {
    annotations: annotations.slice(0, MAX_ANNOTATIONS),
    truncated: true,

    restCount: annotations.length - MAX_ANNOTATIONS,
  };
}

export async function createCheck({
  octokit,
  owner,
  repo,
  name,
  headSha,
}) {
  const res = await octokit.rest.checks.create({
    owner,
    repo,
    name,
    head_sha: headSha,
    status: "in_progress",
  });

  return res.data.id;
}

export async function completeCheck({
  octokit,
  owner,
  repo,
  checkRunId,
  conclusion,
  title,
  summary,
  annotations = [],
}) {
  await octokit.rest.checks.update({
    owner,
    repo,
    check_run_id: checkRunId,
    conclusion,
    status: "completed",
    output: {
      title,
      summary,
      annotations,
    },
  });
}

export async function findExistingCheckRun({
  octokit,
  owner,
  repo,
  name,
  headSha,
}) {
  const res = await octokit.rest.checks.listForRef({
    owner,
    repo,
    ref: headSha,
    check_name: name,
    filter: "latest",
    per_page: 1,
  });

  return res.data.check_runs[0] ?? null;
}

export async function ensureCheckRun({
  octokit,
  owner,
  repo,
  name,
  headSha,
}) {
  const existing = await findExistingCheckRun({
    octokit,
    owner,
    repo,
    name,
    headSha,
  });

  if (existing) {
    return existing.id;
  }

  // 없으면 새로 생성
  const res = await octokit.rest.checks.create({
    owner,
    repo,
    name,
    head_sha: headSha,
    status: "in_progress",
  });

  return res.data.id;
}

export async function createPendingChecks({
  githubToken,
  owner,
  repo,
  headSha,
  checks,
}) {
  const octokit = createOctokit(githubToken);

  await Promise.all(
    checks.map((name) =>
      octokit.rest.checks.create({
        owner,
        repo,
        name,
        head_sha: headSha,
        status: "in_progress",
      })
    )
  );
}
