// step-common/handler.js
import { createOctokit } from "./github.js";
import { ensureCheckRun, completeCheck, splitAnnotations } from "./check.js";

export function withStep({ name, run }) {
  return async function handler(event) {
    const { repository, pullRequest, githubToken } = event;

    if (!githubToken) {
      throw new Error("Missing githubToken in step context");
    }

    // ✅ dispatcher에서 받은 token으로만 Octokit 생성
    const octokit = createOctokit(githubToken);

    // ✅ idempotent Check Run
    const checkRunId = await ensureCheckRun({
      octokit,
      owner: repository.owner,
      repo: repository.name,
      name,
      headSha: pullRequest.headSha,
    });

    try {
      const result = await run({ event, octokit });

      const { annotations, truncated, restCount } =
        splitAnnotations(result.annotations ?? []);

      let summary = result.summary ?? "";
      if (truncated) {
        summary += `\n\n⚠️ 표시되지 않은 annotation ${restCount}개가 더 있습니다.`;
      }

      await completeCheck({
        octokit,
        owner: repository.owner,
        repo: repository.name,
        checkRunId,
        conclusion: result.conclusion ?? "success",
        title: result.title ?? `${name} result`,
        summary,
        annotations,
      });
    } catch (e) {
      await completeCheck({
        octokit,
        owner: repository.owner,
        repo: repository.name,
        checkRunId,
        conclusion: "failure",
        title: `${name} failed`,
        summary: e?.message ?? "unknown error",
      });
    }
  };
}
